Make notification reporting more resilient

luismfonseca opened the pull request:
## Why?

We had a customer report with the following stack trace:
```
Fatal Exception: java.lang.IllegalStateException: bundle.getString(BUNDLE_INSTANCE_ID_KEY) must not be null
       at com.pusher.pushnotifications.reporting.ReportingJobService$Companion.fromBundle(ReportingJobService.java:71)
       at com.pusher.pushnotifications.reporting.ReportingJobService.onStartJob(ReportingJobService.java:100)
       at com.firebase.jobdispatcher.JobService$2.run(JobService.java:164)
       at android.os.Handler.handleCallback(Handler.java:873)
       at android.os.Handler.dispatchMessage(Handler.java:99)
       at android.os.Looper.loop(Looper.java:201)
       at android.app.ActivityThread.main(ActivityThread.java:6810)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:547)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:873)
```

I'm not sure when this can happen, but these operations are best-effort, so we should fail silently if something is occasionally wrong for some reason.

luismfonseca commented on pushnotifications/src/main/java/com/pusher/pushnotifications/reporting/ReportingJobService.kt:
```diff
@@ -63,15 +63,16 @@ class ReportingJobService: JobService() {
-    fun fromBundle(bundle: Bundle): ReportEvent {
-      val event = ReportEventType.valueOf(bundle.getString(BUNDLE_EVENT_TYPE_KEY))
-      when (event) {
+    fun fromBundle(bundle: Bundle): ReportEvent? {
```
The biggest difference is making this return a `ReportEvent?`. So we no longer have the expectation that the bundle will be correctly formed 100% of the times.

marekoid commented on pushnotifications/src/main/java/com/pusher/pushnotifications/reporting/ReportingJobService.kt:
```diff
@@ -63,15 +63,16 @@ class ReportingJobService: JobService() {
           return DeliveryEvent(
-            instanceId = bundle.getString(BUNDLE_INSTANCE_ID_KEY),
-            deviceId = bundle.getString(BUNDLE_DEVICE_ID_KEY),
+            instanceId = bundle.getString(BUNDLE_INSTANCE_ID_KEY) ?: return null,
```
That's won't fix the issue as it's the bundle that is null.

luismfonseca commented:
@marekoid as discussed in person: the assumption on where the null pointer was, was correct but I've added a test to test it explicitly. With regards to the overly-defensive code, I've remove the extraneous `return null` for cases we don't expect and added the global `try .. catch()`.

marekoid commented on pushnotifications/src/main/java/com/pusher/pushnotifications/reporting/ReportingJobService.kt:
```diff
@@ -79,8 +85,14 @@ class ReportingJobService: JobService() {
+          // we are processing a bundle that was created with an old SDK version that
+          // didn't had this key. Our migration strategy is to drop the reporting
+          // as it's (a) a rare one-time transition and (b) it's a best effort feature.
+          val instanceId = bundle.getString(BUNDLE_INSTANCE_ID_KEY) ?: return null
```
Why not to put this line and the comment only once as the first line in this method? It is a common convention to check prerequisites at the beginning of a function. It would be also more legible to have clear if-return statement as opposed to jumping out of the fun from within the elvis operator.

luismfonseca replied:
> Why not to put this line and the comment only once as the first line in this method?
Because we don't know the type before we parse it. And we could have a type that has `instanceId`. Well, probably all of them will have `instanceId`, so I've changed this in d7b6cda.

marekoid commented on pushnotifications/src/main/java/com/pusher/pushnotifications/reporting/ReportingJobService.kt:
```diff
@@ -23,7 +23,7 @@ data class PusherMetadata(
 }
 
-class ReportingJobService: JobService() {
+open class ReportingJobService: JobService() {
```
Production code should not be compromised for the purpose of the test code. Well designed code is naturally testable. When it comes to this class though making it open is not a biggie though but it would be good to at least document why it's marked as such and that actually it's not designed for extension. Even better would be to extract that logic and have a test mostly testing our code as opposed to stretching to testing and relying on deprecated Firebase JobDispatcher and all the other 3rd party dependencies which make the test way more complicated (and more costly in the terms of future maintenance).

marekoid commented on pushnotifications/src/main/java/com/pusher/pushnotifications/reporting/ReportingJobService.kt:
```diff
@@ -63,12 +63,21 @@ class ReportingJobService: JobService() {
+        // it's possible that we are processing a bundle that was created with an old SDK
+        // version that didn't had this key. Our migration strategy is to drop the reporting
+        // as it's (a) a rare one-time transition and (b) it's a best effort feature.
+        return null
```
I think the only change to work around the missing migration should be throwing a specific exception here and catching only that specific exception when calling this method (narrow try-catch).

luismfonseca commented on pushnotifications-integration-tests/src/androidTest/java/com/example/pushnotificationsintegrationtests/ReportingTest.kt:
```diff
@@ -0,0 +1,231 @@
+ * Instrumented test, which will execute on an Android device.
+ *
+ * See [testing documentation](http://d.android.com/tools/testing).
+ */
```
This is the default comments you get when adding a class, I didn't write them but didn't remove them either.

luismfonseca commented on pushnotifications-integration-tests/src/androidTest/java/com/example/pushnotificationsintegrationtests/ReportingTest.kt:
```diff
@@ -0,0 +1,231 @@
+  }
+
+  @Before
+  @After
```
This helps to make sure we don't have any unexpected state. If there's a crash and cleanup doesn't happen, we don't affect other test runs, so as a principle, we should always do both.

luismfonseca commented on pushnotifications-integration-tests/src/androidTest/java/com/example/pushnotificationsintegrationtests/ReportingTest.kt:
```diff
@@ -0,0 +1,231 @@
+      }
+    } catch (e: ConditionTimeoutException) {
+      // Maybe FCM is complaining in CI, so let's pretend to have a token now
+      MessagingService.onRefreshToken!!("fake-fcm-token")
```
So this is the same in all the tests -- we should do a PR to refactor this and move it to a single place as opposed to duplicating it everywhere. :/

luismfonseca commented on pushnotifications-integration-tests/src/androidTest/java/com/example/pushnotificationsintegrationtests/ReportingTest.kt:
```diff
@@ -0,0 +1,231 @@
+
+  @Test
+  @Ignore("The job dispatcher will schedule the job but it isn't triggered within a reasonable timeframe")
+  fun aNotificationSentFromPusherShouldBeReportedTestBroken() {
```
> Not sure if there's any value in adding a broken test which is marked as ignored since day one.

If we don't include this test, one could wonder about why we are not doing a more realistic test and choosing a more hacky way of doing it. They might try to write the test themselves only to realise it's not possible, thus wasting time. By leaving it in, we inform why it's not working (see the `@Ignore` comment) and we leave open the possibility of being fixed later on -- if someone knows how to make it work or if we refactor the internal behaviour.

luismfonseca commented on pushnotifications-integration-tests/src/androidTest/java/com/example/pushnotificationsintegrationtests/ReportingTest.kt:
```diff
@@ -0,0 +1,231 @@
+    bundle.putLong("google.sent_time", 1574938426317)
+    bundle.putLong("google.ttl", 2419200)
+    bundle.putString("google.original_priority", "high")
+    bundle.putString("gcm.notification.e", "1")
```
I don't think it relies on any undocumented internals. It's true that I've grabbed all the properties that a PN had, but it doesn't mean our reporting relies on it.

I wanted to make the test more realistic by not leaving those out (how do we test that we don't fail if they are included?). I can write a comment explaining this in the code to prevent this confusion.

luismfonseca commented on pushnotifications-integration-tests/src/androidTest/java/com/example/pushnotificationsintegrationtests/ReportingTest.kt:
```diff
@@ -0,0 +1,231 @@
+
+    targetCtx.sendBroadcast(i)
+
+    await.atMost(90, TimeUnit.MINUTES) until {
```
No, it is why it's marked as `Ignore`! :D

luismfonseca commented on pushnotifications-integration-tests/src/androidTest/java/com/example/pushnotificationsintegrationtests/ReportingTest.kt:
```diff
@@ -0,0 +1,231 @@
+            .setRetryStrategy(RetryStrategy.DEFAULT_EXPONENTIAL)
+            .setLifetime(Lifetime.UNTIL_NEXT_BOOT)
+            .setExtras(bundledEvent)
+            .build()
```
This test is not the test we should be writing (see previous test).

> Can we have that logic extracted to a separate component in the production

Fair enough.

luismfonseca commented on pushnotifications-integration-tests/src/androidTest/java/com/example/pushnotificationsintegrationtests/ReportingTest.kt:
```diff
@@ -0,0 +1,231 @@
+
+    class ReportingJobServiceWithContext : ReportingJobService() {
+      init {
+        attachBaseContext(context)
```
Me too, but the test compiles, runs and passes. I don't know why the IDE doesn't know about the existence of this function. :/

luismfonseca commented on pushnotifications-integration-tests/src/androidTest/java/com/example/pushnotificationsintegrationtests/ReportingTest.kt:
```diff
@@ -0,0 +1,231 @@
+    val storedDeviceId = getStoredDeviceId()
+
+    // and the device id should exist in the server
+    assertNotNull(errolClient.getDevice(storedDeviceId!!))
```
>  Each test should test a single thing.

Yes, and these assertions are just making sure the pre-conditions are met.

daniellevass commented:
Closing this one out as we've decided to keep #98.
